Evm create join functions

cryptoAtwill opened the pull request:
A lot of the changes are related subnet id update and ROOTNET_ID removal. For review, please focus on the file: `src/manager/evm.rs`, `src/server/handlers/manager/create.rs` and `src/server/handlers/manager/join.rs`.

adlrocha commented on Cargo.toml:
```diff
@@ -82,9 +83,11 @@ fvm_ipld_blockstore = "0.1"
-ipc-sdk = { git = "https://github.com/consensus-shipyard/ipc-actors.git", tag = "v0.2.0" }
-ipc-subnet-actor = { git = "https://github.com/consensus-shipyard/ipc-actors.git", features = [], tag = "v0.2.0" }
-ipc-gateway = { git = "https://github.com/consensus-shipyard/ipc-actors.git", features = [],  tag = "v0.2.0"  }
+
```
You can stop pointing to this branches to avoid having to propagate the changes for the new implementation of the `SubnetID`, that way it will be easier to pick up the changes in the rebase.

cryptoAtwill replied:
So we just use the latest master branch?

adlrocha commented on src/server/handlers/manager/join.rs:
```diff
@@ -52,7 +52,7 @@ impl JsonRPCRequestHandler for JoinSubnetHandler {
 
         let subnet_config = conn.subnet();
-        check_subnet(subnet_config)?;
+        // check_subnet(subnet_config)?;
```
We should still check the parent (I guess you commented it due to the Subnet change)

cryptoAtwill replied:
yeah, I would say the checking needs to be updated as well. EVM and FVM checking should be different.

adlrocha commented on src/manager/evm.rs:
```diff
@@ -26,127 +30,213 @@ pub type MiddlewareImpl = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
+
+        let call = self.registry_contract.new_subnet_actor(params);
+        let pending_tx = call.send().await?;
+        let receipt = pending_tx.retries(10).await?;
```
Let's add a comment explaining that we need to use retries per this issue: https://github.com/filecoin-project/community/discussions/638

adlrocha commented on src/manager/evm.rs:
```diff
@@ -26,127 +30,213 @@ pub type MiddlewareImpl = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
+
+                            log::debug!("subnet with id {subnet_id:?} deployed at {subnet_addr:?}");
+
+                            // subnet_addr.to_string() returns a summary of the actual address, not
```
```suggestion
                            // subnet_addr.to_string() returns a summary of the actual Ethereum address, not
```

adlrocha commented on src/manager/evm.rs:
```diff
@@ -26,127 +30,213 @@ pub type MiddlewareImpl = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
+        _from: Address,
         collateral: TokenAmount,
-        params: JoinParams,
+        _params: JoinParams,
```
There are no params here? Don't we need to pass the multiaddress for the validator as part of the join? If this is the case there's something wrong with the solidity contract

cryptoAtwill replied:
Yeah, the solidity contract does not take the validator address as parameter. Let me check with them to verify this on the latest code.

adlrocha commented on src/manager/evm.rs:
```diff
@@ -26,127 +30,213 @@ pub type MiddlewareImpl = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
+        let mut txn = contract.join();
+        txn.tx.set_value(collateral);
+
+        txn.send().await?.await?;
```
If the transaction doesn't go through this should fail, right?

cryptoAtwill replied:
yep, it will fail.

adlrocha commented on src/manager/evm.rs:
```diff
@@ -159,20 +249,124 @@ impl EthSubnetManager<MiddlewareImpl> {
+                .ok_or_else(|| anyhow!("evm gateway address not defined"))?;
+            let gateway_address = H160::from_str(evm_gateway_address)?;
+
+            let evm_registry_address = subnet
```
Let's call this the IPC registry address throughout the code

adlrocha commented on src/manager/evm.rs:
```diff
@@ -159,20 +249,124 @@ impl EthSubnetManager<MiddlewareImpl> {
     }
 }
+
+// fn evm_id_to_address(evm_subnet: ipc_registry::SubnetID) -> Result<Address> {
```
Remove?

cryptoAtwill replied:
removed

adlrocha commented on src/manager/evm.rs:
```diff
@@ -159,20 +249,124 @@ impl EthSubnetManager<MiddlewareImpl> {
+    }
+}
+
+fn agent_subnet_to_evm_addresses(subnet: &SubnetID) -> Result<Vec<ethers::types::Address>> {
```
Should we make it a proper From implementation? 
```suggestion
fn subnet_to_evm_address(subnet: &SubnetID) -> Result<Vec<ethers::types::Address>> {
```

adlrocha commented on src/manager/evm.rs:
```diff
@@ -159,20 +249,124 @@ impl EthSubnetManager<MiddlewareImpl> {
+//     Ok(Address::from(eth_addr))
+// }
+
+fn agent_subnet_to_evm_address(subnet: &SubnetID) -> Result<ethers::types::Address> {
```
What is the different with the function below? Can you please add comments describing what each function is?

cryptoAtwill replied:
comments added

adlrocha commented on src/manager/evm.rs:
```diff
@@ -159,20 +249,124 @@ impl EthSubnetManager<MiddlewareImpl> {
+    use std::str::FromStr;
+
+    #[test]
+    fn test_addr_convert() {
```
This test has now assert, should we skip it?

cryptoAtwill replied:
removed

adlrocha commented on src/manager/evm.rs:
```diff
@@ -159,20 +249,124 @@ impl EthSubnetManager<MiddlewareImpl> {
+    }
+
+    #[test]
+    fn test_agent_subnet_to_evm_address() {
```
Same here

cryptoAtwill replied:
this one has assert

adlrocha commented on testing/itest/src/infra/mod.rs:
```diff
@@ -300,6 +300,11 @@ impl SubnetInfra {
             accounts,
+            evm: false,
+            eth_private_key: None,
+            chain_id: None,
```
The chainID won't be needed, it can be infered from the subnetID. See: https://github.com/consensus-shipyard/ipc-actors/pull/109/files#diff-41548e2e7ed0210b54962a6cde729ac906ac66516f1c87615c4db169b3230930R71

cryptoAtwill replied:
updated.

adlrocha reviewed (changes requested):
Great job, thanks! A set of suggestions and improvements. Let's fix and clean this one, merge and test the SubnetID refactor, and try to get this one merged so we can move to the implementation of more functions with Solidity.

If you have questions we can go through the review sync.

cryptoAtwill commented:
@adlrocha I have addressed most of the comments above. BTW, probably it's a good idea to refactor `Config` as well, to separate EVM and FVM config so that we can handle the checking and parsing differently. Now they are sharing parameters and it's easy to get confused which is which, especially if we have to support both EVM and FVM simultaneously (not sure if this will happen, but just in case).

cryptoAtwill commented on src/manager/evm.rs:
```diff
@@ -26,127 +30,213 @@ pub type MiddlewareImpl = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
         &self,
         subnet: SubnetID,
-        from: Address,
+        _from: Address,
```
Enum is our friend.
